src/xtgeo/xyz_common/_xyz_data.py

```diff
@@ -0,0 +1,625 @@
+
+                    if self._undef_disc in codes:
+                        del codes[self._undef_disc]
+                    if const.UNDEF_DISC in codes:
+                        del codes[const.UNDEF_DISC]
+
+                    self._attr_records[attr_name] = codes
+
+            # correct when attr_types is CONT but attr_records for that entry is a dict
+            if (
+                attr_name in self._attr_records
+                and self._attr_types[attr_name] == _AttrType.CONT
+            ):
+                if isinstance(self._attr_records[attr_name], dict):
+                    self._attr_records[attr_name] = CONT_DEFAULT_RECORD
+
+    def _ensure_consistency_df_dtypes(self):
+        """Ensure that dataframe float32/64 for all logs, except for XYZ -> float64.
+
+        Whether it is float32 or float64 is set by self._floatbits. Float32 will save
+        memory but loose some precision. For backward compatibility, float64 is default.
+        """
+
+        col = list(self._df)
+        logger.debug("columns: %s", col)
+
+        coords_dtypes = [str(entry) for entry in self._df[col[0:3]].dtypes]
+
+        if not all(["float64" in entry for entry in coords_dtypes]):
+            self._df[col[0:3]] = self._df.iloc[:, 0:3].astype("float64")
+
+        attr_dtypes = [str(entry) for entry in self._df[col[3:]].dtypes]
+
+        if not all([self._floatbits in entry for entry in attr_dtypes]):
+            self._df[col[3:]] = self._df.iloc[:, 3:].astype(self._floatbits)
+
+        for name, attr_type in self._attr_types.items():
+            if attr_type == "CONT":
+                logger.debug("Replacing CONT undef...")
+                self._df[name].replace(
+                    self._undef_cont,
+                    np.float64(const.UNDEF_CONT).astype(self._floatbits),
+                    inplace=True,
+                )
+            else:
+                logger.debug("Replacing INT undef...")
+                self._df[name].replace(
+                    self._undef_disc, np.int32(const.UNDEF_DISC), inplace=True
+                )
+        logger.info("Processed dataframe: %s", list(self._df.dtypes))
+
+    def ensure_consistency(self) -> bool:
+        """Ensure that data and attr* are consistent.
+
+        This is important for many operations on the dataframe, an should keep
+        attr_types and attr_records 'in sync' with the dataframe.
+
+        * When adding one or columns to the dataframe
+        * When removing one or more columns from the dataframe
+        * ...
+
+        Returns True is consistency is ran, while False means that no changes have
+        occured, hence no consistency checks are done
+        """
+
+        # the purpose of this hash check is to avoid psending time on consistency
+        # checks if no changes
+        hash_proposed = (
+            jhash(self._df),
+            jhash(self._attr_types),
+            jhash(self._attr_records),
```

Comment:
I'm mostly just curious, but if you don't mind, why `joblib.hash`?

Reply:
Googled it to be fast and safe hashing for numpy / pandas objects. https://joblib.readthedocs.io/en/latest/generated/joblib.hash.html